### PR TITLE
#2 Display Stored Recommendations

### PR DESCRIPTION
**Pull request by lornaarmstrong**
- Fetch the lat, lng and id for each Recommendation in the datastore 
- Create a marker holding these three values.
- Display marker on the map as a greyIcon

### CODE REVIEW

**Comment by lornaarmstrong** [comment]
Also: removed Marker storage from datastore and instead, only store Recommendations.

File: src/main/java/com/google/sps/servlets/AllMarkersServlet.java [hunk 1]
```diff
@@ -43,17 +43,18 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-      double lat = (double) entity.getProperty("lat");
-      double lng = (double) entity.getProperty("lng");
-      Marker marker = new Marker(lat, lng);
-      allMarkers.add(marker);
-    }
+     /* Collection to hold all markers from datastore */
+    Collection<Marker> allMarkers = new ArrayList<>();
+     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+     Query query = new Query("Recommendation");
+     PreparedQuery results = datastore.prepare(query);
```
Comment (raniarusly) [inline src/main/java/com/google/sps/servlets/AllMarkersServlet.java]:
```suggestion
     Query query = new Query("Recommendation");
     PreparedQuery results = DatastoreServiceFactory.getDatastoreService().prepare(query);
```

File: src/main/java/com/google/sps/servlets/AllMarkersServlet.java [hunk 2]
```diff
@@ -43,17 +43,18 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-    for (Entity entity : results.asIterable()) {
-      double lat = (double) entity.getProperty("lat");
-      double lng = (double) entity.getProperty("lng");
-      Marker marker = new Marker(lat, lng);
-      allMarkers.add(marker);
-    }
+     /* Collection to hold all markers from datastore */
+    Collection<Marker> allMarkers = new ArrayList<>();
+     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+     Query query = new Query("Recommendation");
```
Comment (raniarusly) [inline src/main/java/com/google/sps/servlets/AllMarkersServlet.java]: Since we will only be needing the lat and lng, [projection (https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PropertyProjection)(com.google.appengine.api.datastore.PropertyProjection) might be able to make the interaction with datstore lighter and reduce the memory needed to store the result
Using projection we can get only the necessary attribute, instead of all the attributes.

File: src/main/webapp/scripts/map.js [hunk 3]
```diff
@@ -172,11 +172,11 @@ function placeMarkerAndPanTo(latLng, map) {
   const marker = new google.maps.Marker({
     position: latLng,
     map: map,
-    icon: greyIcon
+    //icon: greyIcon
   });
   /* Recenter the Map */
   map.panTo(latLng);
-  togglePopup(latLng, marker);
+  togglePopup(latLng);
```
Comment (raniarusly) [inline src/main/webapp/scripts/map.js]: I am not sure but togglePopup function below does not take any parameter

File: src/main/webapp/scripts/map.js [hunk 4]
```diff
@@ -172,11 +172,11 @@ function placeMarkerAndPanTo(latLng, map) {
   const marker = new google.maps.Marker({
     position: latLng,
     map: map,
-    icon: greyIcon
+    //icon: greyIcon
```
Comment (raniarusly) [inline src/main/webapp/scripts/map.js]: maybe delete this?

File: src/main/webapp/scripts/map.js [hunk 5]
```diff
@@ -187,20 +187,19 @@ function placeMarkerAndPanTo(latLng, map) {
...
 /* Function to place markers from the datastore */
-function placeMarker(latLng, map) {
+function placeMarker(markerDetails, map) {
   const marker = new google.maps.Marker ({
-    position: latLng,
+    position: new google.maps.LatLng(markerDetails.lat, markerDetails.lng),
     map: map,
     icon: greyIcon, // TODO change the colour of the icon depending on the category
   });
+  marker.set("id", markerDetails.id);
```
Comment (raniarusly) [inline src/main/webapp/scripts/map.js]: are we going to add the "click" eventlistener to the marker for listing the recommendation here?
if yes I was thinking maybe, instead of setting the id,  we can just put the id into the fucntion, e.g. google.maps.event.addListener(marker, "click", () => listRecommendation(markerDetails.id));
listingRecommendation function will fetch the recommendation details from datastore using the Id then proceed to display the details ( line 126-134)
then we can remove addHardCodedMarkers  function

File: src/main/webapp/scripts/map.js [hunk 6]
```diff
@@ -209,24 +208,13 @@ function populateLocation(pos) {
...
-
 /* Fetch all markers from datastore and add to map*/
-function fetchMarkers(map) {
+function fetchMarkerInfo(map) {
   fetch('/all-markers')
   .then(response => response.json())
   .then((markers) => {
     markers.forEach((marker) => {
-      placeMarker(new google.maps.LatLng(marker.lat, marker.lng), map)
+      placeMarker(marker, map)
```
Comment (raniarusly) [inline src/main/webapp/scripts/map.js]:
```suggestion
      placeMarker(marker, map);
```

File: src/main/webapp/scripts/map.js [hunk 7]
```diff
@@ -187,20 +187,19 @@ function placeMarkerAndPanTo(latLng, map) {
...
+    position: new google.maps.LatLng(markerDetails.lat, markerDetails.lng),
     map: map,
     icon: greyIcon, // TODO change the colour of the icon depending on the category
   });
+  marker.set("id", markerDetails.id);
 }
 
 /* Set the PopUp to Active */
-function togglePopup(latLng) {
+function togglePopup() {
```
Comment (raniarusly) [inline src/main/webapp/scripts/map.js:243]: how do we pre-fill the "location" input ? is it somewhere else, not every toggling the popup?

**Review by raniarusly — CHANGES_REQUESTED** [review]
Thank you Lorna, well done! :D

File: src/main/webapp/scripts/map.js (hunk 7 above)
Comment (lornaarmstrong) [reply src/main/webapp/scripts/map.js:243]: it's in the populateLocation function, called within the placeMarkerAndPanTo function - this hasn't changed in many CLs.

File: src/main/webapp/scripts/map.js (hunk 3 above)
Comment (lornaarmstrong) [reply src/main/webapp/scripts/map.js]: Good point - this was updated in a more up-to-date branch but I'll update it here too

File: src/main/webapp/scripts/map.js (hunk 5 above)
Comment (lornaarmstrong) [reply src/main/webapp/scripts/map.js]: Yes, there will be an event listener added but that is not at this iteration of the code - this is just to store and then place the markers (rather than listing).

File: src/main/java/com/google/sps/servlets/AllMarkersServlet.java (hunk 2 above)
Comment (lornaarmstrong) [reply src/main/java/com/google/sps/servlets/AllMarkersServlet.java]: Thanks for the link! This sounds like a great idea.

Comment (lornaarmstrong) [reply src/main/java/com/google/sps/servlets/AllMarkersServlet.java]: As you have already read up on the documentation and have implemented this in the algorithm section, would you be able to commit the code / suggest a change? With the timeframes, I think this might be helpful.

**Comment by lornaarmstrong** [comment]
Thanks for adding the projection @raniarusly! LGTM

File: src/main/webapp/scripts/map.js (hunk 7 above)
Comment (raniarusly) [reply src/main/webapp/scripts/map.js:243]: right, thank you for the clarification :)

**Review by raniarusly — APPROVED** [review]

**Review by MostafaGabriel — COMMENTED** [review]
Thanks for rebasing onto the master branch! Can you please resolve the conflicts too and submit the updated code for review? :)

File: src/main/webapp/recommendation-map.html [hunk 8]
```diff
@@ -49,10 +49,10 @@ <h3>Want to view a recommendation?</h3>
...
       <div class="map-container">
         <div class="map-category-section">
-          <button class="map-category-button" id="restaurants-button">Restaurants</button>
-          <button class="map-category-button" id="places-button">Places to Visit</button>
-          <button class="map-category-button" id="bars-and-clubs-button">Bars and Clubs</button>
-          <button class="map-category-button" id="study-button">Study Places</button>
+          <button class="map-category-button" id="restaurants-button" onclick="initMap('Restaurants')">Restaurants</button>
+          <button class="map-category-button" id="places-button" onclick="initMap('Places To Visit')">Places to Visit</button>
+          <button class="map-category-button" id="bars-and-clubs-button" onclick="initMap('Bars and Clubs')">Bars and Clubs</button>
+          <button class="map-category-button" id="study-button" onclick="initMap('Study Places')">Study Places</button>
```
Comment (MostafaGabriel) [inline src/main/webapp/recommendation-map.html:55]: I really have no idea why this change is included here 🤔 It was added in https://github.com/googleinterns/step241-2020/commit/c1625c8ae4bfb10dae2d93ed3e01ce201445a21c

File: src/main/webapp/scripts/map.js [hunk 9]
```diff
@@ -188,19 +196,54 @@ function placeMarkerAndPanTo(latLng, map) {
...
+
+  /* Add Listener for Click on Marker */
+   marker.addListener("click", () => {
+     fetchRecommendationInfo(marker.id);
+   });
+}
+
+/* Get recommendation from datastore and fill html */
+function fetchRecommendationInfo(id) {
+  const params = new URLSearchParams();
```
Comment (MostafaGabriel) [inline src/main/webapp/scripts/map.js]: Please delete

File: src/main/webapp/scripts/map.js [hunk 10]
```diff
@@ -94,21 +94,29 @@ function initMap() {
...
 
-function addHardCodedMarkers(map) {
-  /* Iterate through stored recommendations to get details */
-  for (i = 0; i < recommendations.length; i++) {
-    const recommendation = recommendations[i];
+/* Add Hard-Coded Markers to map */
+function addMarker(map, recommendationsToAdd) {
+  /* Iterate through stored recommendationsToAdd to get details */
+  for (i = 0; i < recommendationsToAdd.length; i++) {
+    const recommendation = recommendationsToAdd[i];
```
Comment (MostafaGabriel) [inline src/main/webapp/scripts/map.js:119]: ditto, this is already pushed with https://github.com/googleinterns/step241-2020/commit/c1625c8ae4bfb10dae2d93ed3e01ce201445a21c

**Review by MostafaGabriel — COMMENTED** [review]
Thanks Lorna!

File: src/main/webapp/recommendation-map.html (hunk 8 above)
Comment (lornaarmstrong) [reply src/main/webapp/recommendation-map.html:55]: Me neither!

File: src/main/webapp/scripts/map.js (hunk 10 above)
Comment (lornaarmstrong) [reply src/main/webapp/scripts/map.js:119]: Yes, not sure why it's included

File: src/main/webapp/recommendation-map.html (hunk 8 above)
Comment (MostafaGabriel) [reply src/main/webapp/recommendation-map.html:55]: Interesting.. hopefully it will not cause any problem when you merge into master :)

File: src/main/webapp/scripts/map.js (hunk 9 above)
Comment (MostafaGabriel) [reply src/main/webapp/scripts/map.js]: Can you please push that change before merging? :)

**Review by MostafaGabriel — APPROVED** [review]
LGTM! Thanks Lorna :)

File: src/main/webapp/scripts/map.js (hunk 9 above)
Comment (lornaarmstrong) [reply src/main/webapp/scripts/map.js]: Done! Sorry, I must have pushed just after I requested your review :)
